Add/Improve Unit Tests

AliSoftware opened the pull request:
- Add Tests for `listMultipartUploads` & `listParts` Request recently introduced in https://github.com/Automattic/hostmgr/pull/89
- Add some AWSResponses parsing tests (that were missing for existing response types)

AliSoftware committed:
Tests for listMultipartUploads & listParts Request

AliSoftware committed:
Add some response parsing tests

AliSoftware committed:
Make CompleteMultipartUploadRequestBuilder test more resilient

AliSoftware committed:
RFC specifies ETags are expected to contain quotes

Despite this looking odd at first, the literal quotes are supposed to be fully part of the ETag string; in fact, the spec says that an ETag might be prefixed with `W/` before the `quoted-string`, e.g. `W/"a1b2c3"`, so it'd be invalid to remove quotes from that.

This is specified explicitly in https://www.rfc-editor.org/rfc/rfc2616#section-3.11:
```
      entity-tag = [ weak ] opaque-tag
      weak       = "W/"
      opaque-tag = quoted-string

      quoted-string  = ( <"> *(qdtext | quoted-pair ) <"> )
```

So better respect this and keep the literal quotes for consistency with the spec and what the S3 API returns (and expects back)

AliSoftware committed:
There's a Helper for that.

AliSoftware commented on Sources/tinys3/responses/S3HeadResponse.swift:
```diff
@@ -16,7 +16,7 @@ public struct S3HeadResponse {
             let contentLengthString = response.value(forHTTPHeaderField: .contentLength),
             let contentLength = Int(contentLengthString),
-            let eTag = response.value(forHTTPHeaderField: .eTag)?.replacingOccurrences(of: "\"", with: ""),
+            let eTag = response.value(forHTTPHeaderField: .eTag),
```
As [per the HTTP spec/RFC](https://www.rfc-editor.org/rfc/rfc2616#section-3.11), the literal quotes are fully part of the ETag string (in fact, according to the RFC, the quoted string in an ETag _might_ be prefixed with unquoted `W/`, e.g. `W/"a1b2c3"`; so it'd be invalid to remove quotes from such ETag). So this is not a quirk or bug of the S3 API, and instead those quotes are expected and part of the ETag and thus need to be kept.

```
      entity-tag = [ weak ] opaque-tag
      weak       = "W/"
      opaque-tag = quoted-string

      quoted-string  = ( <"> *(qdtext | quoted-pair ) <"> )
```

Keeping the quotes also helps keeping consistency, otherwise only this part of the code would remove the quotes from the ETag, while every other parts of the code dealing with ETags, as well as the S3 API itself, still expects to have those quotes kept verbatim in the ETags.

_In particular, we wouldn't want an `eTag` value previously fetched via a `S3HeadResponse` then later used in a subsequent request (e.g. `S3MultipartUploadCompleteXMLBuilder`, or whatever request includes ETags)… to break that subsequent request just because we trimmed the quotes from the eTag during `S3HeadResponse`._

crazytonyli commented on Tests/tinys3Tests/requests/ListMultipartUploadsRequestTests.swift:
```diff
@@ -0,0 +1,66 @@
+host:examplebucket.s3.amazonaws.com
+x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+x-amz-date:20130524T000000Z
+""", request.canonicalHeaderString)
```
Nitpick: indent the multiline string to be one level deep than `XCTAssertEqual`?

```
swift repl
Welcome to Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4).
Type :help for assistance.
  1> let string = """
  2.     hello
  3.     world
  4.     """
string: String = "hello\nworld"
  5> print(string)
hello
world
```

AliSoftware replied:
Good point. Fixed in all request tests (existing and new) in 342de5de7cf7477ca2f30bd1699049d3c49c434a

crazytonyli commented on Tests/tinys3Tests/requests/ListMultipartUploadsRequestTests.swift:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    func testThatSignatureIsValid() throws {
+        XCTAssertEqual("5feb3c99fc520055ad41dcb6c99dfb714d6bfdf670eb44a4dd14b289b9a5b4ed", request.signature)
```
Is the expected signature text here generated using the `stringToSign` value in the above test and the `request.credentials`?

Is there a way to get the expected signature without running any hostmgr code? Because if for some reason, the `stringToSign` or `credentials` is changed, we'll need to update this signature here, and I guess we should just run the "testee" code and use its result.

Also, consiering the correlation between these two tests, maybe we can merge them into one?

AliSoftware replied:
> Is the expected signature text here generated using the stringToSign value in the above test and the request.credentials?

Yes. In practice `AWSRequest.signature` simply call's its `AWSRequestSigner`'s `.sign(string: stringToSign)`, so this kinda tests the `AWSRequestSigner` more than the `AWSRequest` itself, though it also confirms that the request uses the expected signer.

> Is there a way to get the expected signature without running any hostmgr code? Because if for some reason, the stringToSign or credentials is changed, we'll need to update this signature here, and I guess we should just run the "testee" code and use its result.

Yeah, this is tricky. The only way to ensure this test works is to run it, and if it fails, copy the actual result of the signer as the expected value for the test… so in a way we're writing a test to fit the result instead of the other way around.

There's no real way to get the expected signature without running any `hostmgr` code, as the only way to confirm or deny if our implementation that computes the signature is correct or not is by actually sending a signed request to AWS and see if their server accepts or rejects it. (Btw, this is how I found out that our implementation was incorrect back in https://github.com/Automattic/hostmgr/pull/89#discussion_r1533114140)

While I'm not loving this need to run the test to find out and adjust the expected value—or to send the request to AWS live and check if it rejects it—I'm not sure there's any better way to write a test against the signature algorithm implementation.

> Also, consiering the correlation between these two tests, maybe we can merge them into one?

That could be an option, but this also tests that the `AWSRequest.signature` implementation indeed calls the expected `AWSRequestSigner` implementation (the one implementing AWS4's HMAC algorithm), and that signer uses not only the proper `stringToSign` but also the proper request date and credentials to compute its value. So I think this still has value to keep this as a separate test method?

Besides, all those requests-related tests (`tinys3Tests/requests/*Tests.swift`) conform to our `RequestTest` protocol which expects those test methods 🙃

crazytonyli replied:
I tried to use `aws s3 presign` to generate a signed URL, which is essentially what we are doing here. However, the cli doesn't accept an `AWSRequest.date` equivalent (it probably always uses the "now" timestamp).

```
AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY AWS_DEFAULT_REGION=us-east-1 aws s3 presign s3://examplebucket/images/xcode-14.3.vmtemplate.aar
```

There are AWS SDKs in many different languages. They must have implemented signing requests, where there are more granular control over the parameters. We could potentially add some comment here with ruby/python code that calls AWS SDK with the parameters in `AWSRequest` to get an expected signature.

AliSoftware replied:
`aws s3 presign` is a bit different, in that it implements [the "query string authentication alternative"](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html#RESTAuthenticationQueryStringAuth) approach, as opposed to create a signature for an arbitrary request and provide that signature via `Authorization` header. This also only works for GET requests to access an S3 object—as in [`PresignedURLGetRequestTests.swift`](https://github.com/Automattic/hostmgr/blob/trunk/Tests/tinys3Tests/presign/PresignedURLGetRequestTests.swift)—as it's designed to create URLs that can be shared as direct links (e.g. over Slack or as an HTML link in web page) to S3 objects.

---

So instead, in 555dd80f49b9798d6e3f0be84959416c3a8d5ec7, I've added a `printDebugSigningRubyCode()` method on `AWSRequest` in `TestHelpers.swift` that prints Ruby code to be pasted in your favorite Ruby interpreter (`irb`, `pry`, …) and will print you the expected values for those test methods, as per returned by the official `aws-sigv4` Ruby gem.

So whenever you want to implement a new test, like I did in 7184f82a36dc8bf3728924924b73bd6a8678ca42, you can put a breakpoint in one of the test methods, call `po printDebugSigningRubyCode()`, copy/paste the code from the Console to `irb` to run it, and then copy/paste the answers printed from Ruby back to your Swift code 🎉

crazytonyli replied:
That looks great! Thanks for looking into it 👍

crazytonyli reviewed (commented):
🎉 Yay! More tests!

AliSoftware committed:
Adjust multiline strings indentation in tests

As suggested by @crazytonyli in https://github.com/Automattic/hostmgr/pull/90#discussion_r1536928696

AliSoftware committed:
Add debug ruby code to validate Signing Tests expected values

As discussed in https://github.com/Automattic/hostmgr/pull/90#discussion_r1538534703 with @crazytonyli

AliSoftware committed:
Add Tests for UploadPartRequest

We missed tests for that one too, and that's the only one that has some HTTP Body
